src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -182,32 +193,99 @@ public async Task SendRequestBodyWithExpect100ContinueAsync(CancellationToken ca
                 {
-                    // We received a positive response from the server, or we didn't receive a response but our timer expired.
-                    // In either case, start sending the request body.
-                    await SendRequestBodyAsync(cancellationToken).ConfigureAwait(false);
+                    return;
                 }
-                else
+
+                if (NetEventSource.IsEnabled) Trace($"Stream complete.");
+
+                if (_resetException == null &&
+                    (_requestCompletionState == StreamCompletionState.Failed || _responseCompletionState == StreamCompletionState.Failed))
                 {
-                    // We received a negative response from server, so we will not send the request body, and instead we will reset the stream.
-                    if (NetEventSource.IsEnabled) Trace("Avoiding sending 100-Continue request body content.");
-                    _shouldSendRequestBody = false;
-                    _shouldSendRequestBodyWaiter = null;
                     IgnoreExceptions(_connection.SendRstStreamAsync(_streamId, Http2ProtocolErrorCode.Cancel));
                 }
+
+                _connection.RemoveStream(this);
+
+                // Do cleanup.
+                _streamWindow.Dispose();
+
+                if (_requestBodyCancellationSource != null)
+                {
+                    _requestBodyCancellationSource.Dispose();
+                }
+            }
+
+            private void CancelRequestBody()
+            {
+                CancellationTokenSource requestBodyCancellationSource;
+
+                lock (SyncObject)
+                {
+                    if (_requestCompletionState != StreamCompletionState.InProgress)
+                    {
+                        return;
+                    }
+
+                    requestBodyCancellationSource = _requestBodyCancellationSource;
+                }
+
+                if (NetEventSource.IsEnabled) Trace($"Canceling sending request body.");
+                requestBodyCancellationSource.Cancel();
+
+                // When cancellation propagates, SendRequestBodyAsync will set _responseCompletionState to Failed
+            }
+
+            public void CancelResponseBody()
+            {
+                bool signalWaiter;
+
+                lock (SyncObject)
+                {
+                    if (_responseCompletionState != StreamCompletionState.InProgress)
+                    {
+                        return;
+                    }
+
+                    _responseCompletionState = StreamCompletionState.Failed;
+                    CheckForCompletion();
+
+                    _responseProtocolState = ResponseProtocolState.Aborted;
+
+                    signalWaiter = _hasWaiter;
+                    _hasWaiter = false;
+                }
+
+                if (signalWaiter)
+                {
+                    _waitSource.SetResult(true);
+                }
```

Comment:
The thread-safety of _waitSource is described at:
https://github.com/dotnet/corefx/blob/e64cac6dcacf996f98f0b3f75fb7ad0c12f588f7/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L50-L61

Does all of that still hold with your changes?  I've not seen anything yet to suggest otherwise, but just want to make sure.

Reply:
Yes.